libpoolprotocols/stratum/EthStratumClient.cpp

```diff
@@ -340,56 +343,62 @@ void EthStratumClient::connect_handler(const boost::system::error_code& ec, tcp:
 				// Do not trigger a full disconnection but, instead, let the loop
 				// continue with another IP (if any). 
 				// Disconnection is triggered on no more IP available
-
+				m_connected.store(false, std::memory_order_relaxed);
 				m_socket->close();
 				start_connect(++i);
 				return;
 			}
 		}
 
-		// This should be done *after* a valid connection which may fail
-		// on secure connection.
-		m_connected.store(true, std::memory_order_relaxed);
+		// Trigger event handlers and begin counting for the next job
 		if (m_onConnected) { m_onConnected(); }
-
-		// Successfully connected so we start our work timeout timer
 		reset_work_timeout();
 
 		string user;
 		size_t p;
 
+		Json::Value jReq;
+		jReq["id"] = unsigned(1);
+		jReq["method"] = "mining.subscribe";
+		jReq["params"] = Json::Value(Json::arrayValue);
+
+		p = m_conn.User().find_first_of(".");
+		user = m_conn.User().substr(0, p);
+		if (p + 1 <= m_conn.User().length())
+			m_worker = m_conn.User().substr(p + 1);
+		else
+			// Some marketing ?
+			m_worker = ("ethminer " + toString(ethminer_get_buildinfo()->project_version));
+
 		switch (m_conn.Version()) {
 
 			case EthStratumClient::STRATUM:
 
-				sendSocketData("{\"id\": 1, \"method\": \"mining.subscribe\", \"params\": []}");
+				jReq["jsonrpc"] = "2.0";
+
 				break;
 
 			case EthStratumClient::ETHPROXY:
 
-				p = m_conn.User().find_first_of(".");
-				user = m_conn.User().substr(0, p);
-				if (p + 1 <= m_conn.User().length())
-					m_worker = m_conn.User().substr(p + 1);
-				else
-					m_worker = "";
 
-				if (m_email.empty())
-				{
-					sendSocketData("{\"id\": 1, \"worker\":\"" + m_worker + "\", \"method\": \"eth_submitLogin\", \"params\": [\"" + user + "\"]}");
-				}
-				else
-				{
-					sendSocketData("{\"id\": 1, \"worker\":\"" + m_worker + "\", \"method\": \"eth_submitLogin\", \"params\": [\"" + user + "\", \"" + m_email + "\"]}");
-				}
+				jReq["method"] = "eth_submitLogin";
+				jReq["worker"] = m_worker;
+				jReq["params"].append(user);
+				if (!m_email.empty()) jReq["params"].append(m_email);
+
 				break;
 
 			case EthStratumClient::ETHEREUMSTRATUM:
 
-				sendSocketData( "{\"id\": 1, \"method\": \"mining.subscribe\", \"params\": [\"ethminer/" + toString(ethminer_get_buildinfo()->project_version) + "\",\"EthereumStratum/1.0.0\"]}");
+				jReq["params"].append("ethminer " + toString(ethminer_get_buildinfo()->project_version));
```

Comment:
Do not use `toString()`, it's enough to construct a string from `const char*`. `"ethminer " + std::string{ethminer_get_buildinfo()->project_version}`

Reply:
Ok